# **Práctica 1: Aritmética de punto flotante** 
## Parte computacional
Alumnos:

*   Olvera Trejo Alberto
*   Ramos López Gabriela



## Ejercicio 1
El número de Euler se puede definir como:
## $$ e =\underset{n\rightarrow\infty}{\lim} \left(1 + \dfrac{1}{n}\right)^n$$</font>
Escribe un programa que, a manera de tabla arroje los valores de $(1 + 1/n)^n$ para $n = 10^k$, con $k = 0,1,2,...,20$, y en cada paso calcule los errores absolutos y relativos comparando con el valor de $e$ dado por la máquina. Utiliza doble precisión de IEEE para hacer los cálculos.

In [ ]:
from math import e
import pandas as pd
import decimal
import math
import numpy as np

#definimos una funcion que calcule el limite
def exp(n):
  return((1+ (1/n))**n)

#creamos las listas que van a tener los valores absolutos y realtivos
error_absoluto = []
error_relativo = []
exponente = []

#hacemos las 20 iteraciones y guardamos los resultado
for i in range(20):
  aprox = exp(10**i)
  error_r = abs(e - aprox)
  error_a = error_r/abs(e)

  #usamos la libreria decimal para guardar todos los decimales
  error_absoluto.append(decimal.Decimal.from_float(error_a))
  error_relativo.append(decimal.Decimal.from_float(error_r))
  exponente.append('10^' + str(i+1))

#Creamos un dataframe para mostrar los resultados
errores = pd.DataFrame({'Valor de n' : exponente, 'Error relativo': error_relativo, 'Error absoluto' : error_absoluto})

errores

Valor de n                                     Error relativo  \
0        10^1  0.71828182845904509079559829842764884233474731...   
1        10^2  0.12453936835904277913300575164612382650375366...   
2        10^3  0.01346799903751660920647736929822713136672973...   
3        10^4  0.00135789622345150462479068664833903312683105...   
4        10^5  0.00013590163411958400274670566432178020477294...   
5        10^6  0.00001359126674760346986658987589180469512939...   
6        10^7  0.00000135936329170860403792175929993391036987...   
7        10^8  1.34326963330266835328075103461742401123046875E-7   
8        10^9     3.01116873657747419201768934726715087890625E-8   
9       10^10     2.23552515166147713898681104183197021484375E-7   
10      10^11  2.24775742463378946922603063285350799560546875E-7   
11      10^12  2.24898065059875307269976474344730377197265625E-7   
12      10^13  0.00024166757819266138085367856547236442565917...   
13      10^14  0.00217179437214420900659206381533294916152954...   
14      10^15  0.00217179437202208447388329659588634967803955...   
15      10^16  0.31675337809021675283815966395195573568344116...   
16      10^17  1.71828182845904509079559829842764884233474731...   
17      10^18  1.71828182845904509079559829842764884233474731...   
18      10^19  1.71828182845904509079559829842764884233474731...   
19      10^20  1.71828182845904509079559829842764884233474731...   

                                       Error absoluto  
0   0.26424111765711533195144511410035192966461181...  
1   0.04581547323576906638464834031765349209308624...  
2   0.00495459995961913379075580721178084786515682...  
3   0.00049954210385215155106747308977332977519836...  
4   0.00004999541721419838180041905761896714466274...  
5   0.00000499994761592037140142570639245178654164...  
6   5.00081808102733617769019257293106051065478823...  
7   4.94161282041953890180591808193449887554038468...  
8   1.10774707208503926517693924966127816400529582...  
9   8.22403743517928146628743796449356739941549676...  
10  8.26903745263238946084093398611325476110778254...  
11  8.27353744947656156024870261617154554301123425...  
12  0.00008890453361477210778365809495937810424948...  
13  0.00079895849996369512190791395411793018865864...  
14  0.00079895849991876795228445962138152935949619...  
15  0.11652705572099553155140938542899675667285919...  
16  0.63212055882855766597572255705017596483230590...  
17  0.63212055882855766597572255705017596483230590...  
18  0.63212055882855766597572255705017596483230590...  
19  0.63212055882855766597572255705017596483230590...

In [ ]:
print(len(str(decimal.Decimal.from_float(exp(10**14)))))

Decimal('0.002171794372144209006592063815332949161529541015625')

> - <font color=red> ¿Qué observas con las aproximaciónes de $e$ y los errores?</font>

Conforme el exponente de $10^n$ aumenta el error se va reduciendo hasta que llega a $n = 13$ que es cuando el error aumenta, es decir que el error de la aproximación $n = 13$ es mayor que el de la aproximación $n = 14$
> - <font color=red> Explica lo que sucede.</font>

Lo anterior se puede explicar ya que cuando el exponente llega a $n = 13$ la aproximación tiene una longitud de 51 dígitos, lo cual es casi la misma cantidad de bits disponibles para la mantisa. Luego, cuando $n = 14$ la aproximación supera el número de bits disponibles para la mantisa, con lo cual se empieza a truncar/redondear por lo que vamos a generar errores en el proceso, lo cual explica lo sucedido

## Ejercicio 2
La función exponencial se puede aproximar con su serie de Taylor alrededor de cero: 

## $$e^{x} = 1 + x + \frac{x^{2}}{2} + \cdots + \frac{x^{n}}{n!} + \cdots $$

Escribe un programa que calcule el valor aproximado de $e^x$ para cualquier $x$ mediante los primeros $n$ términos de su serie de Taylor. Como criterio o condición de paro propón que $\frac{x^n}{n!}<10^{-6}$. El programa debe evitar cancelaciones para valores negativos de x.

In [ ]:
def exp_pos(x, n):
  exp = 0

  for i in range (n):
      if (x**i)/math.factorial(i) < 10**(-6):
        break
        
      exp += decimal.Decimal.from_float(((x**i)/math.factorial(i)))
  return exp

def expo(x, n):
  if  x > 0:
    return exp_pos(x, n)
  else:
    return(1/(exp_pos(-x, n)))

## Ejercicio 4
Escribe un programa que en un ciclo, en cada iteración, calcule el valor asignado a una variable y se sustituya por el doble del valor anterior, es decir, $x_{n+1}=2x_n$. Inicia con $x_0=1$ y utilice como condición de paro que $x_{n+1} \leq x_n$. Haz los cálculos en formato de precisión sencilla de IEEE. Imprime el número de iteraciones y los cálculos en forma de tabla.

In [ ]:
valores = [-20, -15, -10, -5, -1, 1, 5, 10, 15, 20]
valores_aprox = []
valores_real = []

for num in valores:
  aprox = expo(num, 100)
  real = decimal.Decimal.from_float(math.exp(num))

  valores_aprox.append(aprox)
  valores_real.append(real)

tabla = pd.DataFrame({'x':valores, 'Serie de Taylor':valores_aprox, 'Valor Real':valores_real})

tabla

x                     Serie de Taylor  \
0 -20    2.061153622438560512518693188E-9   
1 -15    3.059023205019066373262583775E-7   
2 -10  0.00004539992976451083727726286854   
3  -5    0.006737947025459595223227753734   
4  -1      0.3678794821625895941365380235   
5   1       2.718281525573192228231432792   
6   5       148.4131585216477746468045717   
7  10       22026.46579382377874874633369   
8  15       3269017.372471246649001667955   
9  20       485165195.4097896460649407729   

                                          Valor Real  
0  2.06115362243855786994151713966121758359051341...  
1  3.05902320501825785895684203796740519010199932...  
2  0.00004539992976248485417314557088808157914172...  
3  0.00673794699908546700084510661099557182751595...  
4  0.36787944117144233402427744294982403516769409...  
5  2.71828182845904509079559829842764884233474731...  
6  148.413159102576599934764089994132518768310546875  
7            22026.465794806717894971370697021484375  
8            3269017.3724721106700599193572998046875  
9                   485165195.4097902774810791015625

> - <font color=red> Genera una tabla que muestre las aproximaciones que se obtienen con tu algoritmo para $x = \pm20,\pm15,\pm10,\pm5,\pm1$ y que también muestre las evaluaciones hechas por la función exponecial dada por la máquina.

## Ejercicio 3

Investiga como obtener la épsilon de máquina $\varepsilon$, en doble precisión, de Python. Implementa lo anterior en un programa e imprime el valor de $\varepsilon$, además calcula la siguiente cantidad:

## $$\epsilon = \vert 3(4/3 -1) -1 \vert$$

Imprime el valor de $\epsilon$.

In [ ]:
import numpy as np

abs_eps = abs(3*(4/3-1)-1)

def epsilonMaq(func=float):
  maq_epsilon = func(1)
  while func(1)+func(maq_epsilon) != func(1):
    maq_epsilon_n = maq_epsilon
    maq_epsilon = func(maq_epsilon) / func(2)
  return maq_epsilon_n

eps = np.finfo(np.float64).eps

emaq = epsilonMaq(np.float64)

print("ϵ = |3(4/3-1)-1| = ",abs_eps, " - - operación dada")
print("ε = ", emaq , " - - función implementada para calcular la épsilon de máquina")
print("ε_2 = ",eps, " - - épsilon calculada con finfo de numpy")

ϵ = |3(4/3-1)-1| =  2.220446049250313e-16  - - operación dada
ε =  2.220446049250313e-16  - - función implementada para calcular la épsilon de máquina
ε_2 =  2.220446049250313e-16  - - épsilon calculada con finfo de numpy


> - <font color=red> Explica porque $\epsilon = \varepsilon$</font>

Por definición sabemos que la épsilon de la máquina es el número decimal más pequeño que sumado a 1 nos arroja un valor mayor a 1. Vemos que $\epsilon$ es tan pequeña como $\varepsilon$ ya que al redondear $(4/3)-1 = 1/3 = 0.33\bar{3}$) y entonces tendremos que $\epsilon = \vert 3(0.33̄\bar{3}) -1 \vert = \vert 0.99\bar{9} - 1 \vert$ obteniendo así un número tan pequeño que cumple ser $ɛ$ ya que sabemos que 1/3 es "periodico" en nuestro sistema de doble precisón, así podemos asegurar que existirá un "residuo".

> - <font color=red> ¿El cálculo anterior funcionaría para un sistema de punto flotante de base 3?</font>

No funcionaría, ya que 1/3 en un sistema de punto flotante de base 3 tendría una representadión finita, y no existiría este "residuo" pequeño

In [ ]:
x = 1
iteracion = 0
historial = []

while x < np.float32(2*x):
    x = np.float32(2*x)
    iteracion += 1
    historial.append(x)

In [ ]:
print(historial[-3:])
print(iteracion)
print(len(historial[:-1]))

[8.507059e+37, 1.7014118e+38, inf]
128
127


> - <font color=red> En artimética real la condición de paro nunca se cumple, explique porque sí se cumple para el formato de precisión sencilla de IEEE.</font>

Se cumple porque al seguir multiplicando $x$ por dos, el número es cada vez más grande hasta llegar a un overflow, es decir, no hay suficiente precisión para poder representar dicho número en la máquina

> - <font color=red> ¿Cuántas iteraciones realizó el ciclo?</font>

Se realizaron 128 iteraciones
> - <font color=red> ¿Cuáles fueron el último y penultimo valor de la variable?</font>

Antes de obtener el overflow, los números calculados fueron:
$$8.507059\times 10^{37} \hspace{2em} 1.7014118 \times 10^{38}$$